Add `fraction` argument support for export dataloader

ambitious-octopus opened the pull request:
<!--
Thank you 🙏 for your contribution to [Ultralytics](https://www.ultralytics.com/) 🚀! Your effort in enhancing our repositories is greatly appreciated. To streamline the process and assist us in integrating your Pull Request (PR) effectively, please follow these steps:

1. Check for Existing Contributions: Before submitting, kindly explore existing PRs to ensure your contribution is unique and complementary.
2. Link Related Issues: If your PR addresses an open issue, please link it in your submission. This helps us better understand the context and impact of your contribution.
3. Elaborate Your Changes: Clearly articulate the purpose of your PR. Whether it's a bug fix or a new feature, a detailed description aids in a smoother integration process.
4. Ultralytics Contributor License Agreement (CLA): To uphold the quality and integrity of our project, we require all contributors to sign the CLA. Please confirm your agreement by commenting below:

    I have read the CLA Document and I sign the CLA

For more detailed guidance and best practices on contributing, refer to our ✅ [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Your adherence to these guidelines ensures a faster and more effective review process.
-->


## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Added a new `fraction` argument to support dataset fraction selection for INT8 quantization during model export.

### 📊 Key Changes
- Introduced a `fraction` argument for export formats like OpenVINO, TensorRT, TFLite, and IMX.
- Updated documentation to reflect the new `fraction` argument across relevant export formats.
- Enhanced the exporter to utilize the `fraction` parameter for INT8 calibration, allowing users to specify a subset of the dataset for calibration.

### 🎯 Purpose & Impact
- **Purpose**: To provide users with greater flexibility in INT8 quantization by enabling calibration on a smaller dataset subset, saving time and computational resources.
- **Impact**: This is particularly useful for resource-constrained environments or quick experimentation, improving usability for edge and lightweight deployments. 🚀

ambitious-octopus committed:
added fraction support for export

UltralyticsAssistant commented:
👋 Hello @ambitious-octopus, thank you for submitting an `ultralytics/ultralytics` 🚀 PR! To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Your PR description looks great! Ensure all key details, such as the purpose, impact, and relevant issues, are clearly outlined. If applicable, link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues) for added context.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues to ensure compatibility.
- ✅ **Update Documentation**: If applicable, update the relevant [documentation](https://docs.ultralytics.com/) to reflect your changes, particularly for the new `fraction` parameter.
- ✅ **Add Tests**: Include or update tests to cover the new `fraction` parameter functionality. Confirm that all tests are passing to maintain code integrity.
- ✅ **Sign the CLA**: If this is your first contribution, please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to only what's necessary for the feature addition. _"Simplicity is the ultimate sophistication."_ — Leonardo da Vinci

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don’t hesitate to leave a comment if you have any questions.

This is an automated response, but an Ultralytics engineer will review your PR and provide additional feedback soon. Thank you for contributing to Ultralytics! 🌟🚀

lakshanthad committed:
Add `fraction` argument to all export docs

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

lakshanthad commented:
Thanks! @ambitious-octopus . I have tested this and it works! I have also updated all the export docs pages to reflect this change. @glenn-jocher FYI.

lakshanthad committed:
Merge branch 'main' into dataset-fraction-export-support

glenn-jocher committed:
Merge branch 'main' into dataset-fraction-export-support

glenn-jocher commented:
@ambitious-octopus @lakshanthad we have an allowed arguments table that @Y-T-G implemented in Exporter, **any** argument added to Exporter must be added to this table or it will be rejected.

https://github.com/ultralytics/ultralytics/blob/bf6602af9800f0da7d2f73f97bab7240c5abc68b/ultralytics/engine/exporter.py#L135-L156

glenn-jocher reviewed (changes requested):
Argument must be added to export arguments verification function otherwise this PR serves no purpose and the argument will be unusable.

ambitious-octopus committed:
added export validation

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

glenn-jocher committed:
Merge branch 'main' into dataset-fraction-export-support

ambitious-octopus committed:
Merge branch 'main' into dataset-fraction-export-support

glenn-jocher committed:
Update export-table.md

@ambitious-octopus you forgot your new macro Export-table here!

Signed-off-by: Glenn Jocher <glenn.jocher@ultralytics.com>

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

glenn-jocher committed:
Update tutorial.ipynb

@ambitious-octopus @lakshanthad @ultralytics/yolo please **don't forget to update the tutorial.ipynb export table** as well when modifying any export arguments.

Signed-off-by: Glenn Jocher <glenn.jocher@ultralytics.com>

glenn-jocher committed:
Created using Colab

glenn-jocher commented:
@ambitious-octopus @lakshanthad I've updated tables that were missing this new argument. Please remember to always modify all relevant places when updating the repo:

- https://github.com/ultralytics/ultralytics/blob/main/docs/en/macros/export-table.md
- https://github.com/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb

glenn-jocher commented:
This is a great addition that will help users optimize their INT8 calibration process, especially with larger datasets. The `fraction` parameter significantly reduces export time when calibrating with just a representative subset of data (like using 0.1 for 10% of the calibration dataset).

For users who want to try this: when exporting with INT8 quantization, you can now specify what portion of the calibration dataset to use:

```python
# Export with INT8 using only 20% of the calibration dataset
model.export(format="engine", int8=True, fraction=0.2)
```

This will be particularly valuable for edge device deployments where finding the right balance between calibration accuracy and processing time is crucial.
